src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -1632,67 +1108,36 @@ private void SetRequestHandleTlsOptions(SafeWinHttpHandle requestHandle)
 
         private void SetRequestHandleClientCertificateOptions(SafeWinHttpHandle requestHandle, Uri requestUri)
         {
-            // Must be HTTPS scheme to use client certificates.
-            if (requestUri.Scheme != UriSchemeHttps)
+            if (requestUri.Scheme != WinHttpRequestState.UriSchemeHttps)
             {
                 return;
             }
 
-            // Get candidate list for client certificates.
-            X509Certificate2Collection certs;
+            X509Certificate2 clientCertificate = null;
             if (_clientCertificateOption == ClientCertificateOption.Manual)
             {
-                certs = ClientCertificates;
+                clientCertificate = WinHttpCertificateHelper.GetEligibleClientCertificate(ClientCertificates);
             }
             else
             {
-                using (var myStore = new X509Store())
-                {
-                    myStore.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
-                    certs = myStore.Certificates;
-                }
-            }
-
-            // Check for no certs now as a performance optimization.
-            if (certs.Count == 0)
-            {
-                SetNoClientCertificate(requestHandle);
-                return;
+                clientCertificate = WinHttpCertificateHelper.GetEligibleClientCertificate();
             }
 
-            // Reduce the set of certificates to match the proper 'Client Authentication' criteria.
-            certs = certs.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, false);
-            certs = certs.Find(X509FindType.FindByApplicationPolicy, ClientAuthenticationOID, false);
-
-            // Build a new collection with certs that have a private key. Need to do this
-            // manually because there is no X509FindType to match this criteria.
-            var clientCerts = new X509Certificate2Collection();
-            foreach (var cert in certs)
-            {
-                if (cert.HasPrivateKey)
-                {
-                    clientCerts.Add(cert);
-                }
-            }
-
-            // TOOD: Filter the list based on TrustedIssuerList info from WINHTTP.
-
-            // Set the client certificate.
-            if (clientCerts.Count == 0)
-            {
-                SetNoClientCertificate(requestHandle);
-            }
-            else
+            if (clientCertificate != null)
             {
                 SetWinHttpOption(
                     requestHandle,
                     Interop.WinHttp.WINHTTP_OPTION_CLIENT_CERT_CONTEXT,
-                    clientCerts[0].Handle,
+                    clientCertificate.Handle,
```

Comment:
@bartonjs There is an interesting note regarding not closing the store using `CERT_CLOSE_STORE_FORCE_FLAG` since WinHttp will duplicate the context:
https://msdn.microsoft.com/en-us/library/windows/desktop/aa384066(v=vs.85).aspx

Can you confirm that the X509Store won't do that?

@davidsh We probably need an integration test to ensure that's not going to be added in the future.